src/transport_mmap.c

```diff
@@ -0,0 +1,295 @@
+/* -*- C -*-
+ *
+ * Copyright 2011 Sandia Corporation. Under the terms of Contract
+ * DE-AC04-94AL85000 with Sandia Corporation, the U.S.  Government
+ * retains certain rights in this software.
+ *
+ * Copyright (c) 2023 Intel Corporation. All rights reserved.
+ * This software is available to you under the BSD license.
+ *
+ * This file is part of the Sandia OpenSHMEM software package. For license
+ * information, see the LICENSE file in the top level directory of the
+ * distribution.
+ *
+ */
+
+#include "config.h"
+
+#include <errno.h>
+#include <string.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+
+#define SHMEM_INTERNAL_INCLUDE
+#include "shmem.h"
+#include "shmem_internal.h"
+#include "shmem_comm.h"
+#include "runtime.h"
+#include "transport_mmap.h"
+
+#define MPIDI_OFI_SHMGR_NAME_MAXLEN (128)
+#define MPIDI_OFI_SHMGR_NAME_PREFIX "/sos_shm_mmap_area"
+
+
+static void shm_create_key(char *key, size_t max_size, unsigned pe, size_t num) {
+    snprintf(key, max_size, "%s-%u-%zu", MPIDI_OFI_SHMGR_NAME_PREFIX, pe, num);
+}
+
+
+static void *shm_create_region(char* base, const char *key, int shm_size) {
+  if (shm_size == 0) return NULL;
+
+  shm_unlink(key);
+  int fd = shm_open(key, O_RDWR | O_CREAT | O_TRUNC, 0666);
+  if (fd == -1) {
+      fprintf(stderr, "mmap_init error shm_open with errno(%s)\n", strerror(errno));
+      exit(0);
+  }
+
+  if (ftruncate(fd, shm_size) == -1) {
+      fprintf(stderr, "mmap_init error ftruncate\n");
+      exit(0);
+  }
+
+  void *shm_base_addr = mmap(base, shm_size, PROT_READ | PROT_WRITE, MAP_SHARED | MAP_FIXED, fd, 0);
+  if (MAP_FAILED == shm_base_addr) {
+      fprintf(stderr, "mmap_init error mmap %s size %d\n", key, shm_size);
+      exit(0);
+  }
+
+  return shm_base_addr;
+}
+
+
+static void *shm_create_region_data_seg(char* base, const char *key, int shm_size) {
+    if (shm_size == 0) return NULL;
+
+    shm_unlink(key);
+    int fd = shm_open(key, O_RDWR | O_CREAT | O_TRUNC, 0666);
+    if (fd == -1) {
+        fprintf(stderr, "mmap_init data_seg error shm_open with errno(%s)\n", strerror(errno));
+        exit(1);
+    }
+
+    /* Write all current contents of the data segment to the file */
+    FILE *fp = fdopen(fd, "wb");
+    size_t ret = fwrite(base, shm_size, 1, fp);
```

Comment:
Can you share a reference to docs for this? As far as I've been able to track down, `fdopen` on a descriptor returned by `shm_open` is described as having undefined behavior.

Reply:
I can't find anything referencing this either, probably because it's a complete hack ;).

BTW, there's a good chance we won't merge this PR, at least as it is, for this very reason.  Maybe it should just be a topic branch until the symmetric data segment is deprecated or something.  The reason this WIP PR exists is because I'd like to evaluate some internal systems where there's no chance XPMEM will ever be installed, so we have to think out of the box.

Yes, manpages say `fdopen` on a shared memory object is undefined, but I think that's because there are many operations one may not be able to do on shared memory "files" (like seek, lock via fcntl, chmod, etc.) and possibly because there could be other processes writing to the shared region while the file is being opened.  So it's usually a very bad idea to do this.

But I'm wondering if it may be relatively harmless to do this within SOS init on Linux, because we're not doing any fancy ops on the file (just writing to it and then truncating it) and no other PE or thread will write to the region until after finalize (but I think there's a risk of say some crazy interrupt or something leading to changes to the data segment before it's mmap'd).

I'm surprised this approach always seems to work on my end (based on the unit tests) - but I've only tried a few flavors of Linux.  MacOS definitely doesn't like it at all, I get a failure on `fwrite`.

Reply:
Yeah, this is the only mention I could find: https://stackoverflow.com/questions/23347877/map-char-pointer-to-in-memory-file-with-file-descriptor